### ISSUE TITLE
How do we destroy a namespace?

### ISSUE THREAD

**Issue by diranged** [SUSE/Portus#767]
Lets say we've created a test namespace and now we want to destroy it... how do we do that? The UI certainly doesn't allow it. I know I also need to go and purge the images from the repo, but I'm more concerned with the Portus side of things first.

**Comment by mssola** [comment]
We cannot do that for now. This is something that comes from the registry, in which it does not allow to remove images for now. Since we cannot remove images, then we cannot remove namespaces either, and the thing also goes up because it means that we cannot remove teams either.

We could allow removing namespaces if there are no images there, and the same for teams. But I fear that doing that wouldn't be enough right ?

**Comment by diranged** [comment]
Hrmm -- I was under the impression the more recent releases of Distribution supported deleting image references from the API. Is that not actually true yet?

**Comment by cyphar** [comment]
@diranged They support "soft deletes", where the layers are essentially just untagged (the data is not removed). Currently, there is no support for garbage collection so I would consider the "soft deletion" an **unsafe** feature at the moment.

**Comment by holgerreif** [comment]
I wouldn't call it an _unsafe feature_. As long as documentation clearly states, that only the tag is removed, but the image layers remain, I can't see a big problem. I am even more concernd with a clear external structure than with wasted storage space (which can't be reclaimed anyway right now).

See my [Comment in #641](https://github.com/SUSE/Portus/issues/641#issuecomment-170052971)

**Comment by diranged** [comment]
I agree with @holgerreif, we don't care about the wasted layer space .. we care about being able to clean up our structure when necessary. I would like to see both the "soft delete from repo" option in the Portus UI (just for simplicity) as well as ultimately the "delete unused namespace" from Portus functionality.

**Comment by holgerreif** [comment]
Without touching the registry itsself the namespace could just get the state `abandoned`(see [#751](https://github.com/SUSE/Portus/issues/751#issuecomment-187724647)) inside portus. Thus the namespace wouldn't pop up again once crono's catalog API job runs.

**Comment by mssola** [comment]
As @cyphar explained, there is now just support for soft deletes. This is not something that I'm interesting in doing by itself, because I don't want to lie to users about deleting blobs.

That being said, GC support in the registry has just been merged :D (see [PR](https://github.com/docker/distribution/pull/1386)). The main issue about that is #197. Once that is fixed, we can start implementing deleting namespaces and all that :)

Therefore, I'll keep this issue open as a reminder to extend the removing action to namespaces and teams too.

**Comment by holgerreif** [comment]
That's great news! I was monitoring [docker/distribution#462](https://github.com/docker/distribution/issues/462) where no progess happend...

Then I aggree, you should not implement this feature half way!

OTOH there are several issues discussing what and how to delete, neither of which show the full picture. Should we open a new issue just for a complete picture? Maybe I could come up with a proposal summerizing the ideas in various areas?

**Comment by mssola** [comment]
@holgerreif that would be amazing ! :) I'm currently still investigating on all the nuts & bolts of this newly merged GC support. Maybe there are some gotchas that we need to be aware of.

**Comment by eburgueno** [comment]
I have just hit this issue today, after deleting a bunch of images from our registry (using https://github.com/burnettk/delete-docker-registry-image)

I tried just deleting a row in the `namespaces` table in the MySQL database, but that causes issues with the site when trying to access the admin dashboard (although `admin/namespaces` loads fine). I can't find any references to that namespace nor the namespace id on the database itself, so I'm assuming it's coming from somewhere else?

**Comment by mssola** [comment]
@eburgueno since deleting images is not supported in Portus, this is going to be tricky :) If you want to work-around this, be aware of the `activities` table. This table will contain references to pushed images, created namespaces, etc. Does this help ?

**Comment by eburgueno** [comment]
@mssola Yes. I totally overlooked that table. Took me a minute to figure out that `namespaces.id == activities.trackable_id`.

Deleting the rows with matching IDs for the namespace I wanted to clear seems to have done the trick. I am not sure if this would bring any issues down the line though.

**Comment by rennhak** [comment]
+1

**Comment by gentoid** [comment]
+1

**Comment by sergioasantiago** [comment]
+1

**Comment by raghu999** [comment]
+1

**Comment by crusadergo** [comment]
+1

**Comment by dmigous** [comment]
+1

**Comment by cedvan** [comment]
:+1:

**Comment by mssola** [comment]
Unblocking this one because #197 is no longer a blocking issue :tada:

**Comment by EugenMayer** [comment]
also looking forward to this one. Since registry:2.4 supports deletes and this already has been implemented in portus, namespace deletion should be no big issue either. Great!

**Comment by emcniece** [comment]
@mssola is this delayed again?

**Comment by mssola** [comment]
@emcniece I've changed the name of the milestone so it's less confusing. As described in [this small update](http://port.us.org/2016/10/18/portus-2.1.1-and-planning-for-2.2.html), we want a more clear stance and what we want for the next release (note that for the 2.1 release we wanted to implement an unrealistic number of features/enhancements). So we are kind of doing the contrary now: we want to guarantee a small set of "must" features for the next release, and if we have the time we'll tackle other issues like this one (e.g. this one specifically is something that I really want to see for 2.2).

**Comment by holgerreif** [comment]
Be aware when fixing #899 it must be taken into account, that a namespace can not only get renamed but disapear as well.

**Comment by fhaefemeier** [comment]
Since a few days I started using Portus to give it a try. I am searching a useful UI for a docker registry. Portus seems to fulfill my needs. I found this (long running) issue, because I missed the feature to delete a previous (manually!) created namespace. After reading all comments I can follow the arguments why it isn't yet implemented on one side. But on the other I can't...
Let me explain! I deployed Portus on my machine (as docker container) and after a few tries it was running and conected to my registry. I was wondering why nothing is shown in the frontend, because my registry (provided by a Nexus service) contains several images. I learned I have to create a namespace, which match the same name in my registry and (magically :-)) all related images with there tags appear in Portus. I was happy.
In my understanding there was a loose coupling between a namespace and there counterpart in the registry. Only in case I create manually the namespace in Portus the user can see the images.
This is fine for me.
So my question and the reason why I can't understand all the arguments in the previous comments, why can't I delete a manual created namespace? In my domain (or view of the world, in context of Portus; see previous sentence) if I delete a namespace all related images will disappear from the UI, but not necessarily deleted in the registry. This would be a nice feature but isn't the main one. The same from the other side. If I delete all images of a namespace in the registry (outside), the namespace in Portus should not disappear but empty.
What do you think? Did I miss something?
